File: api/orbita_api/routers/comparisons.py
"""Сравнение запусков и рекомендация (`05_API.md` §2, ADR-006)."""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from orbita_api.adapters.registry import StorageRegistry, get_storage_registry
from orbita_api.db.session import get_session
from orbita_api.schemas.errors import error_responses
from orbita_api.schemas.results import ComparisonRequest, ComparisonResult, Recommendation
from orbita_api.services import comparisons

router = APIRouter(tags=["comparisons"])

Session = Annotated[AsyncSession, Depends(get_session)]
Storage = Annotated[StorageRegistry, Depends(get_storage_registry)]


@router.post(
    "/comparisons",
    summary="Метрики запусков рядом, дельты и изменённые параметры",
    responses=error_responses(400, 404, 503),
)
async def compare_runs(
    request: ComparisonRequest,
    session: Session,
    storage: Storage,
) -> ComparisonResult:
    """Сравнивает запуски на одной сетке времени.

    Базой считается первый запуск списка: дельты должны отсчитываться от того варианта,
    с которым инженер сравнивает, а не от лучшего по метрикам.
    """
    return await comparisons.compare(session, storage, request.run_ids)


@router.get(
    "/runs/{run_id}/recommendation",
    summary="Рекомендация по конфигурации с обоснованием",
    responses=error_responses(400, 404, 503),
)
async def get_run_recommendation(
    run_id: UUID,
    base_run_id: Annotated[
        UUID,
        Query(description="Запуск, относительно которого считаются дельты"),
    ],
    session: Session,
) -> Recommendation:
    """Строит рекомендацию детерминированно из сохранённых метрик (ADR-006, ADR-012)."""
    return await comparisons.recommend(session, run_id, base_run_id)

File: web/src/api/comparisons.ts
import { apiPost, apiRequest } from './client';
import { runEvidencePackPath } from './runs';
import type { ComparisonResult, Recommendation } from './types';

/**
 * `POST /api/comparisons` — метрики запусков рядом, дельты и изменённые параметры.
 * Первый идентификатор списка сервис считает базой (`05_API.md` §1).
 */
export function compareRuns(runIds: readonly string[]): Promise<ComparisonResult> {
  return apiPost<ComparisonResult>('/api/comparisons', { run_ids: [...runIds] });
}

/** `GET /api/runs/{id}/recommendation?base_run_id=` — вывод из ранжирования ADR-006. */
export function getRecommendation(runId: string, baseRunId: string): Promise<Recommendation> {
  return apiRequest<Recommendation>(
    `/api/runs/${runId}/recommendation?base_run_id=${baseRunId}`,
  );
}

/**
 * Путь Evidence Pack со сравнением: без `base_run_id` архив собирается без сравнения и
 * вывода (`05_API.md` §2), а на экране сравнения нужен именно он.
 */
export function comparisonEvidencePackPath(runId: string, baseRunId: string): string {
  return `${runEvidencePackPath(runId)}?base_run_id=${baseRunId}`;
}
